eeps/eep-00XX.md

```diff
@@ -0,0 +1,317 @@
+   and `MFArgs` as arguments;
+
+5. As we attempt to statically type Erlang programs, MFArgs offer
+   limited opportunities for static verification, which either
+   becomes the source of dynamism (so errors that could be caught
+   statically must now be handled at runtime) or leads to false
+   positives (requiring developers to rewrite their code);
+
+Solution
+========
+
+Erlang must provide a contruct for partially applied functions.
+Partially applied functions use the `fun Mod:Fun(...Args)` notation,
+where arguments can also be placeholders given by the `_` variable.
+
+We will break down the syntax in the following section. For now,
+let's see an example:
+
+    1> Fun = fun maps:get(username, _).
+    2> Fun(#{username => "Joe"}).
+    "Joe"
+
+The above is equivalent to:
+
+    1> Fun = fun(X) -> maps:get(username, X) end.
+    2> Fun(#{username => "Joe"}).
+    "Joe"
+
+While the proposed notation does provide syntactical affordances,
+the most important aspect is that the function preserves its remote
+name and arguments within the runtime. This means the partially
+applied function can be passed across nodes or written to disk,
+even if the module that defines the function is gone.
+
+Furthermore, partially applied functions can replace `MFArgs`,
+removing all ambiguity about its behaviour. For example, imagine
+the configuration below:
+
+```erlang
+{some_config, {some_mod, some_fun, [answer, 42]}}.
+```
+
+If `some_config` is invoked with an additional argument, such
+argument is not specified in the configuration definition itself,
+therefore it is unclear which arity of `some_mod:some_fun` will
+be invoked. But with partially applied functions, the number of
+arguments is always clear, "go to definition" works, as config
+files and static typing:
+
+```erlang
+{some_config, fun some_mod:some_fun(_, answer, 42)}.
+```
+
+In practice, they solve all the downsides of `MFArgs` listed above:
+
+1. The arity is always clear;
+
+2. `xref` or "go to definition" can be unambiguously implemented;
+
+3. It is possible to handle different arities via
+   `is_function(Fun, Arity)` checks;
+
+4. There is no longer a need for MFArgs, functions are all you need;
+
+5. They can be statically checked;
+
+Syntax Specification
+====================
+
+The syntax of partially applied functions will be:
+
+```erlang
+fun Fun(...Args)
+fun Mod:Fun(...Args)
+```
+
+Where `Args` can be zero, one, or many arguments. Arguments
+must be either literals or bound variables. We have seen
+literal examples above, but the arguments may also be variables:
+
+```erlang
+Key = get_key().
+Fun = fun maps:get(Key, _).
+```
+
+The `_` variable denotes placeholders, which are arguments
+that have not yet been provided (the use of `_` is a proposal,
+the exact notation can be changed). The number of placeholders
+dictate the arity of the function and they are provided in order.
+For example:
+
+```erlang
+fun hello(_, world, _)
```

Comment:
would partial application in steps be interesting or just confusing?
Fun1 = fun my_foo( _, b, _).
Fun2 = fun Fun1(_, c),
Res = Fun2(a),  %% fun my_foo(a,b,c)

Reply:
@frazze-jobb given `fun Foo/2` is not valid today, I believe we could allow such without introducing any ambiguity, but it does add to the concerns described in the "visual cluttering" section.

Reply:
Actually I believe that right now the EEP does not address the case of `fun Foo(_, c)` where `Foo` is a variable. It does say that this doesn't work fo call a local function if `Foo` is an atom, but it doesn't specify if this should fail to be an function application if `Foo` is bound to another fun.
If this indeed is valid and `Foo` is bound to a "serialisable" fun, should an expression like this produce a "serialisable" fun as well?

Reply:
Yeah, that's my read of the situation too. I will add it later to the EEP as a possible extension.